evennia/plugin.py

```diff
@@ -0,0 +1,217 @@
+"""
+This provides the base class that all Evennia Plugins must implement to be loadable.
+
+NOTE: Do not add any global imports to this file. It must be totally standalone. The launcher and server both
+must be able to import things from it cleanly.
+"""
+
+
+class EvPlugin:
+    """
+    Base class to use for Evennia Plugins.
+
+    NOTE: the constructor and init_settings methods must not import anything or couple itself to code that's
+        server-only. Keep in mind that the Evennia Launcher must also be capable of running __init__ and
+        init_settings() !
+    """
+
+    def name(self) -> str:
+        """
+        The name of the plugin. This is how Evennia and other Plugins will refer to it.
+
+        Returns:
+            name (str)
+        """
+        return f"{self.__class__.__module__}.{self.__class__.__name__}"
+
+    def load_priority(self) -> int:
+        """
+        The order that plugin's load hooks will be called in. the lower the number, the sooner
+        it will be called.
+
+        Returns:
+            number
+        """
+        return 0
+
+    def init_settings(self, settings, plugins):
+        """
+        First hook to be called. Override this to make a plugin alter the Evennia settings.py
+        during the boot-up process.
+
+        WARNING: This is called by the launcher and possibly in other places, and the run state
+            of Evennia is not guaranteed. ONLY use this method to alter the settings.
+
+        Args:
+            settings (settings.py): The module reference to settings.py.
+            plugins (dict of name->EvPlugin classes): Allows plugins to know what
+                plugins will be loading.
+
+        Returns:
+            None
+        """
+
+    def at_plugin_setup(self):
+        """
+        First phase of plugin setup at the end of evennia._init()
+        """
+
+    def at_plugin_collaborate(self):
+        """
+        Second phase of plugin setup during evennia._init(). At this point, all plugins
+        are available in the evennia.PLUGINS dictionary.
+
+        Use this method for any plugin setup that involves plugins collaborating with each other.
+        """
+
+    def at_plugin_finalize(self):
```

Comment:
Not sure this needs to be separated from `at_plugin_collaborate` ... especially since it's a bit unclear that this is not called until after _all_ plugins have been initialized.